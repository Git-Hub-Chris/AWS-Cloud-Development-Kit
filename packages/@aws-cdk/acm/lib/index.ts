export * from './certificate';
export * from './certificate-ref';

// The L1 library for AWS::CertificateManager:
export * from '../cfn/certificatemanager';
