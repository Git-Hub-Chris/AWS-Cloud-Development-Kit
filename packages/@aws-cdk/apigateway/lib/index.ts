// The L1 Library for AWS::ApiGateway:
export * from '../cfn/apigateway';
