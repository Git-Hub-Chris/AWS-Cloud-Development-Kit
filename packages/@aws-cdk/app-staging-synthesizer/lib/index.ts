export * from './default-staging-stack';
export * from './synthesizer';
