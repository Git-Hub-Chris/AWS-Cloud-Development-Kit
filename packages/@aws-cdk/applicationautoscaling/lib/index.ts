// The L1 Library for AWS::ApplicationAutoScaling:
export * from '../cfn/applicationautoscaling';
