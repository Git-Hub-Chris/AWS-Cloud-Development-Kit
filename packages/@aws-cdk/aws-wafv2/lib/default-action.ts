import { Construct } from 'constructs';
import { CfnWebACL } from './wafv2.generated';

/**
 * The type returned from the `bind()` method in {@link DefaultAction}.
 */
export interface DefaultActionConfig {
  /**
   * The configuration for this default action.
   */
  readonly configuration: CfnWebACL.DefaultActionProperty;
}

/**
 * The action to perform if none of the Rules contained in the WebACL match.
 */
export abstract class DefaultAction {
  /**
   * Specifies that AWS WAF should allow requests by default.
   */
  public static allow(): DefaultAction {
    return new AllowAction();
  }

  /**
   * Specifies that AWS WAF should block requests by default.
   */
  public static block(): DefaultAction {
    return new BlockAction();
  }

  /**
   * Returns the DefaultAction configuration.
   */
  public abstract bind(scope: Construct): DefaultActionConfig
}

class AllowAction extends DefaultAction {
  bind(_scope: Construct) {
    return {
      configuration: {
        allow: {},
      },
    };
  }
};

class BlockAction extends DefaultAction {
  bind(_scope: Construct) {
    return {
      configuration: {
        block: {},
      },
    };
  }
};
