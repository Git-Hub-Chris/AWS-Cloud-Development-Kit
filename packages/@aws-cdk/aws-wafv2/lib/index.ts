export * from './default-action';
export * from './web-acl';

// AWS::WAFv2 CloudFormation Resources:
export * from './wafv2.generated';
