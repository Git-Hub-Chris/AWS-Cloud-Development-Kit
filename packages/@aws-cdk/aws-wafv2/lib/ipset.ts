import * as cdk from '@aws-cdk/core';
import * as waf from './index';

/**
 * Specifies whether the addresses are IPV4 or IPV6.
 */
export enum IpAddressVersion {
  /**
   * Specifies IPV4 addresses
   */
  IPV4 = 'IPV4',

  /**
   * Specifies IPV6 addresses
   */
  IPV6 = 'IPV6'
}

/**
 * Specifies whether this is for an AWS CloudFront distribution or for a
 * regional application. A regional application can be an Application Load
 * Balancer (ALB) or an API Gateway stage. Valid Values are CLOUDFRONT and
 * REGIONAL.
 */
export enum IpSetScope {
  /**
   * Specifies a Cloudfront application scope
   */
  CLOUDFRONT = 'CLOUDFRONT',

  /**
   * Specifies a Regional application scope
   */
  REGIONAL = 'REGIONAL'
}

export interface IIpSet {
  /**
   * A friendly name of the IP set. You cannot change the name of an IPSet
   * after you create it.
   */
  readonly ipSetName: string;
}

export abstract class IpSetBase extends cdk.Resource implements IIpSet {
  /**
   * A friendly name of the IP set. You cannot change the name of an IPSet
   * after you create it.
   */
  public abstract readonly ipSetName: string;
}

export interface IpSetProps {
  /**
   * A friendly name of the IP set. You cannot change the name of an IPSet
   * after you create it.
   * @default none
   */
  readonly name?: string;

  /**
   * A friendly description of the IP set. You cannot change the description
   * of an IP set after you create it.
   * @default none
   */
  readonly description?: string;

  /**
   * Contains an array of strings that specify one or more IP addresses or
   * blocks of IP addresses in Classless Inter-Domain Routing (CIDR) notation.
   * AWS WAF supports all address ranges for IP versions IPv4 and IPv6.
   */
  readonly addresses?: string[];

  /**
   * Specify IPV4 or IPV6.
   *
   * @default IPV4
   */
  readonly ipAddressVersion?: IpAddressVersion;

  /**
   * Specifies whether this is for an AWS CloudFront distribution or for a
   * regional application. A regional application can be an Application
   * Load Balancer (ALB) or an API Gateway stage.
   *
   * @default CLOUDFRONT
   */
  readonly scope?: IpSetScope;
}

export class IpSet extends IpSetBase {
  /**
   * A friendly name of the IP set. You cannot change the name of an IPSet
   * after you create it.
   */
  public readonly ipSetName: string;

  /**
   * Contains an array of strings that specify one or more IP addresses or
   * blocks of IP addresses in Classless Inter-Domain Routing (CIDR) notation.
   * AWS WAF supports all address ranges for IP versions IPv4 and IPv6.
   */
  private addresses: string[] = [];

  constructor(scope: cdk.Construct, id: string, props: IpSetProps) {
    super(scope, id, {
      physicalName: props.name || cdk.PhysicalName.GENERATE_IF_NEEDED
    });

    this.addAddresses(...props.addresses || []);

    const resource = new waf.CfnIPSet(this, 'Resource', {
      addresses: {
        ipAddresses: cdk.Lazy.listValue({ produce: () => this.addresses }, { omitEmpty: true })
      },
      ipAddressVersion: props.ipAddressVersion || IpAddressVersion.IPV4,
      name: props.name,
      scope: props.scope || IpSetScope.CLOUDFRONT,
      description: props.description
    });

    this.ipSetName = this.getResourceNameAttribute(resource.ref);
  }

  /**
   * Adds an ip address to the the addresses list
   */
  public addAddresses(...addresses: string[]): void {
    for (const address of addresses) {
      this.addresses.push(address);
    }
  }
}
