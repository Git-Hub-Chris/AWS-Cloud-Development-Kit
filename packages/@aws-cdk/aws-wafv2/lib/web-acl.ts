import { Resource, Names } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { DefaultAction } from './default-action';
import { CfnWebACL } from './wafv2.generated';

/**
 * Specifies whether this is for an Amazon CloudFront distribution or for a regional application.
 * A regional application can be an Application Load Balancer (ALB), an Amazon API Gateway REST API,
 * or an AWS AppSync GraphQL API.
 */
export enum Scope {
  /**
   * For regional application
   */
  REGIONAL = 'REGIONAL',

  /**
   * For Amazon CloudFront distribution
   */
  CLOUDFRONT = 'CLOUDFRONT',
}

/**
 * Defines and enables Amazon CloudWatch metrics and web request sample collection.
 */
export interface VisibilityConfig {
  /**
   * A boolean indicating whether the associated resource sends metrics to Amazon CloudWatch.
   */
  readonly cloudWatchMetricsEnabled: boolean;

  /**
   * The descriptive name of the Amazon CloudWatch metric. The name can contain only alphanumeric characters
   * (A-Z, a-z, 0-9), with length from one to 128 characters. It can't contain whitespace or metric names reserved
   * for AWS WAF, for example "All" and "Default_Action." You can't change a MetricName after you create a VisibilityConfig.
   */
  readonly metricName: string;

  /**
   * A boolean indicating whether AWS WAF should store a sampling of the web requests that match the rules.
   * You can view the sampled requests through the AWS WAF console.
   */
  readonly sampledRequestsEnabled: boolean;
}

/**
 * Properties for defining an AWS WAF web ACL
 */
export interface WebAclProps {
  /**
   * The descriptive name of the web ACL. You cannot change the name of a web ACL after you create it.
   * @default None
   */
  readonly webAclName?: string;

  /**
   * Specifies whether this is for an Amazon CloudFront distribution or for a regional application.
   */
  readonly scope: Scope;

  /**
   * The action to perform if none of the Rules contained in the WebACL match.
   */
  readonly defaultAction: DefaultAction;

  /**
   * Defines and enables Amazon CloudWatch metrics and web request sample collection.
   * @default Set false to both Amazon CloudWatch metrics and web request sample collection
   */
  readonly visibilityConfig?: VisibilityConfig;
}

/**
 * Defines an AWS WAF web ACL in this stack.
 */
export class WebAcl extends Resource {
  /**
   * Name of this web ACL rule
   * @attribute
   */
  public readonly webAclName: string;

  /**
   * The Amazon Resource Name (ARN) of the web ACL.
   * @attribute
   */
  public readonly webAclArn: string;

  /**
   * The current web ACL capacity (WCU) usage by the web ACL.
   * @attribute
   */
  public readonly webAclCapacity: number;

  /**
   * The ID of the web ACL.
   * @attribute
   */
  public readonly webAclId: string;

  /**
   * The label namespace prefix for this web ACL. All labels added by rules in this web ACL have this prefix.
   * @attribute
   */
  public readonly webAclLabelNamespace: string;

  constructor(scope: Construct, id: string, props: WebAclProps) {
    super(scope, id, {
      physicalName: props.webAclName,
    });

    const visibilityConfig: VisibilityConfig = props.visibilityConfig ?? {
      cloudWatchMetricsEnabled: true,
      metricName: props.webAclName || Names.uniqueId(this),
      sampledRequestsEnabled: true,
    };

    const resource = new CfnWebACL(this, 'Resource', {
      name: this.physicalName,
      scope: props.scope,
      defaultAction: props.defaultAction.bind(this).configuration,
      visibilityConfig,
    });

    this.webAclName = this.getResourceNameAttribute(resource.ref);
    this.webAclArn = this.getResourceArnAttribute(resource.attrArn, {
      service: 'wafv2',
      resource: 'webacl',
      resourceName: this.physicalName,
    });
    this.webAclCapacity = resource.attrCapacity;
    this.webAclId = resource.attrId;
    this.webAclLabelNamespace = resource.attrLabelNamespace;
  }
}
