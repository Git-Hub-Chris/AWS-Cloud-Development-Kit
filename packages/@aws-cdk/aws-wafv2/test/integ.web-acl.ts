import * as cdk from '@aws-cdk/core';
import * as wafv2 from '../lib';

const app = new cdk.App();

class TestStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new wafv2.WebAcl(this, 'WebAcl', {
      webAclName: 'my-web-acl-name',
      scope: wafv2.Scope.CLOUDFRONT,
      defaultAction: wafv2.DefaultAction.block(),
      visibilityConfig: {
        cloudWatchMetricsEnabled: false,
        metricName: 'test-metric-name',
        sampledRequestsEnabled: false,
      },
    });
  }
}

new TestStack(app, 'test-stack', {
  env: {
    region: 'us-east-1',
  },
});
app.synth();
