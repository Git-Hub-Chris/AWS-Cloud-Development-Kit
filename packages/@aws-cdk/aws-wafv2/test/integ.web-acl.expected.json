{
  "Resources": {
    "WebAclE76B067C": {
      "Type": "AWS::WAFv2::WebACL",
      "Properties": {
        "DefaultAction": {
          "Block": {}
        },
        "Scope": "CLOUDFRONT",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "my-web-acl-name",
          "SampledRequestsEnabled": true
        },
        "Name": "my-web-acl-name"
      }
    }
  }
}